### PR TITLE
GroupByTargetWrapper - Allow one to group by request-identifier

### PR DESCRIPTION
**Pull request by snakefoot**
Allow one to to take an entire buffer of logevents and group-by request-identifier. Sending individual emails / reports.

### CODE REVIEW

File: tests/NLog.UnitTests/Targets/Wrappers/GroupByTargetWrapperTests.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    using NLog.Targets;
+    using NLog.Targets.Wrappers;
+    using Xunit;
+
+    public class GroupByTargetWrapperTests : NLogTestBase
+	{
+        [Fact]
+        public void SimpleGroupByTest()
+        {
+            // Stage
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Targets/Wrappers/GroupByTargetWrapperTests.cs]: Arrange? ;)

**Comment by snakefoot** [comment]
Still pondering on this one. Whether the GroupBy-logic should "just" be added to https://github.com/NLog/NLog/wiki/PostFilteringWrapper-target

**Comment by 304NotModified** [comment]
Cool! I like this feature!

> Still pondering on this one. Whether the GroupBy-logic should "just" be added to https://github.com/NLog/NLog/wiki/PostFilteringWrapper-target

I think separate would be better findable. What would be the benefit of combining with PostFilteringWrapper?

**Comment by snakefoot** [comment]
I was thinking the following setup:

- AutoFlushWrapper (Delayed flush on error)
  - BufferingWrapper (Overflow = Discard)
    - PostFilteringWrapper (Only write groups that contains one or more errors)
       - MailTarget (Receives batch of messages for each group)

This would allow one to get (almost) all logevents (with same correlationid) for a requests that resulted in an error-message.

**Comment by snakefoot** [comment]
> What would be the benefit of combining with PostFilteringWrapper?

I guess it is only the allocation issue that really matters. And I guess for most people this is a non-issue, since it only happens on error-reporting.

**Comment by 304NotModified** [comment]
> I guess it is only the reduced allocation that is a benefit But I guess for most people this is a non-issue, since it only happens on error-reporting.

agree on that. I think it's more clear to have this one separate. I'm also OK to merge it for 4.x

**Comment by snakefoot** [comment]
Created wiki-page: https://github.com/NLog/NLog/wiki/GroupByWrapper-target

**Pull request merged**